Sources/ClientRuntime/Networking/Http/URI.swift

```diff
@@ -0,0 +1,148 @@
+    public let path: String
+    public let host: String
+    public let port: Int16
+    public let query: [SDKURLQueryItem]
+    public let username: String?
+    public let password: String?
+    public var url: URL? {
+        self.toBuilder().getUrl()
+    }
+    public var queryString: String? {
+        if self.query.isEmpty {
+            return nil
+        }
+
+        return self.query.map { queryItem in
+            return [queryItem.name, queryItem.value].compactMap { $0 }.joined(separator: "=")
+        }.joined(separator: "&")
+    }
+
+    fileprivate init(scheme: Scheme,
+                     path: String,
+                     host: String,
+                     port: Int16,
+                     query: [SDKURLQueryItem],
+                     username: String? = nil,
+                     password: String? = nil) {
+        self.scheme = scheme
+        self.path = path
+        self.host = host
+        self.port = port
+        self.query = query
+        self.username = username
+        self.password = password
+    }
+
+    public func toBuilder() -> URIBuilder {
+        return URIBuilder()
+           .withScheme(self.scheme)
+           .withPath(self.path)
+           .withHost(self.host)
+           .withPort(self.port)
+           .withQueryItems(self.query)
+    }
+}
+
+public class URIBuilder {
+    var urlComponents: URLComponents
+    var scheme: Scheme = Scheme.https
+    var path: String = "/"
+    var host: String = ""
+    var port: Int16 = Int16(Scheme.https.port)
+    var query: [SDKURLQueryItem] = []
+    var username: String?
+    var password: String?
+
+    required public init() {
+        self.urlComponents = URLComponents()
+        self.urlComponents.scheme = self.scheme.rawValue
+        self.urlComponents.path = self.path
+        self.urlComponents.host = self.host
+    }
+
+    @discardableResult
+    public func withScheme(_ value: Scheme?) -> URIBuilder {
+        self.scheme = value ?? Scheme.https
+        self.urlComponents.scheme = self.scheme.rawValue
+        return self
+    }
+
+    @discardableResult
+    public func withPath(_ value: String) -> URIBuilder {
+        self.path = value
+        if self.path.contains("%") {
+            self.urlComponents.percentEncodedPath = self.path
+        } else {
+            self.urlComponents.path = self.path
+        }
+        return self
+    }
+
+    @discardableResult
+    public func withHost(_ value: String) -> URIBuilder {
+        self.host = value
+        self.urlComponents.host = self.host
+        return self
+    }
+
+    @discardableResult
+    public func withPort(_ value: Int16) -> URIBuilder {
+        self.port = value
+        return self
+    }
+
+    @discardableResult
+    public func withQueryItems(_ value: [SDKURLQueryItem]) -> URIBuilder {
+        self.query.append(contentsOf: value)
+        if !self.query.isEmpty {
+            self.urlComponents.percentEncodedQuery = self.query.map { queryItem in
```

Comment:
Why not set this when `query` is empty? Also, do we know that `query` will be properly encoded (since we're assigning to `percentEncodedQuery`)? Rather than build the query string, you could map `SDKURLQueryItem` to `Foundation.URLQueryItem` - they're the same.

Reply:
The url will be `http://example.amazonaws.com?` if an empty query is set.

Reply:
Oh yea I see `URL` and `URLComponents` don't strip the `?` (or omit the port, as I see from your other [comment](https://github.com/smithy-lang/smithy-swift/pull/719#discussion_r1600803964)). How important is the string representation though? I mean if it is that important, we couldn't omit the port anyways, because then it wouldn't be an accurate URI. And for `queryItems` here, they get defaulted to an empty array when constructing `URI` anyways.